### PR TITLE
Make `DualArray` more reasonable

### PR DESCRIPTION
**Pull request by YingboMa**

### CODE REVIEW

File: src/dualnumber.jl [hunk 1]
```diff
@@ -161,6 +159,8 @@ Base.convert(::Type{Dual{T,V,P}}, d::Dual{T}) where {T,V,P} = Dual{T}(convert(V,
 Base.convert(::Type{Dual{T,V,P}}, x) where {T,V,P} = Dual{T}(convert(V, x), zero(P))
 Base.convert(::Type{Dual{T,V,P}}, x::Number) where {T,V,P} = Dual{T}(convert(V, x), zero(P))
 Base.convert(::Type{D}, d::D) where {D<:Dual} = d
+Base.float(d::Dual{T,V,P}) where {T,V,P} = convert(Dual{T,float(V),P}, d)
+Base.AbstractFloat(d::Dual) = float(d)
```
Comment (oxinabox) [inline src/dualnumber.jl]: What's this line for?
Does it match to base behavior? e.g. on Complex?

**Comment by YingboMa** [comment]
If someone wants to play with this branch. Here is my testing script
```julia
using Cthulhu
using BenchmarkTools
using ForwardDiff2: D, ForwardDiff2, dualrun
using OrdinaryDiffEq, ForwardDiff
using LinearAlgebra: norm

function lorenz(du, u, p, t)
    du[1] = 10.0 * (u[2] - u[1])
    du[2] = u[1] * (28.0 - u[3]) - u[2]
    du[3] = u[1] * u[2] - (8 / 3) * u[3]
    return nothing
end
function odesolver(u0, value, tf)
    tspan = (0.0, tf)
    prob = ODEProblem(lorenz, u0, tspan)
    sol = solve(prob, Tsit5(), internalnorm=(u, t) -> value(norm(u)))
    return sol[end]
end

u0 = [1.0, 0.0, 0.0]
foo(u0) = D(x->odesolver(x, ForwardDiff2.value, 10.0))(u0)
foo(u0)

###
### broadcast perf
###
__a = Ref{Any}()
try
D(u0->__a[] = init(ODEProblem(lorenz, u0, (0, 1.0)), Tsit5(), adaptive=false, dt=0.001))(u0)
catch
end
integ = __a[]
du, x, y = map(_->copy(integ.u), 1:3)
foo!(du, a, x, y) = dualrun(()->@. du .= muladd(a, x, y))
@btime foo!($du, 1.0, $x, $y)
@descend foo!(du, 1.0, x, y)
```
With `@descend`, one would see
```julia
CodeInfo(
1 ── %1  = %new(var"#15#16"{ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1},Float64,ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1},ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1}}, du, a, x, y)::var"#15#16"{ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1},Float64,ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1},ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1}}
│    %2  = Base.sle_int(1, 1)::Bool
└───       goto #3 if not %2
2 ── %4  = Base.sle_int(1, 0)::Bool
└───       goto #4
3 ──       nothing::Nothing
4 ┄─ %7  = φ (#2 => %4, #3 => false)::Bool
└───       goto #6 if not %7
5 ──       invoke Base.getindex(()::Tuple, 1::Int64)::Union{}
└───       $(Expr(:unreachable))::Union{}
6 ┄─       goto #7
7 ──       goto #8
8 ──       goto #9
9 ──       goto #10
10 ─       goto #11
11 ─       goto #12
12 ─       goto #13
13 ─       goto #14
14 ─ %19 = invoke ForwardDiff2.overdub($(QuoteNode(Cassette.Context{nametype(DualContext),Nothing,Nothing,ForwardDiff2.var"##PassType#402",Nothing,Cassette.DisableHooks}(nametype(DualContext)(), nothing, nothing, ForwardDiff2.var"##PassType#402"(), nothing, Cassette.DisableHooks())))::Cassette.Context{nametype(DualContext),Nothing,Nothing,ForwardDiff2.var"##PassType#402",Nothing,Cassette.DisableHooks}, %1::var"#15#16"{ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1},Float64,ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1},ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1}})::ForwardDiff2.DualArray{ForwardDiff2.Tag{Nothing},3,Array{Float64,1},Array{Float64,2},ForwardDiff2.Dual{ForwardDiff2.Tag{Nothing},Float64,StaticArrays.SArray{Tuple{3},Float64,1,3}},1}
└───       goto #15
15 ─       return %19
)
Select a call to descend into or ↩ to ascend. [q]uit.
Toggles: [o]ptimize, [w]arn, [d]ebuginfo, [s]yntax highlight for LLVM/Native.
Show: [L]LVM IR, [N]ative code
Advanced: dump [P]arams cache.

 • %9  = invoke getindex(::Tuple,::Int64)::Union{}
   %19  = invoke overdub(::Cassette.Context{…},::var"#15#16"{…})
   ↩
```
I suspect that `%19  = invoke overdub(::Cassette.Context{…},::var"#15#16"{…})` is where the allocation from.

File: src/dualnumber.jl (hunk 1 above)
Comment (YingboMa) [reply src/dualnumber.jl]: https://github.com/YingboMa/ForwardDiff2.jl/pull/19/commits/58ededfcdf7af5e01a6446960d16704351286020

Comment (YingboMa) [reply src/dualnumber.jl]: We don't care about Complex right now.

Comment (oxinabox) [reply src/dualnumber.jl]: The problem is about API.
Generally when `Foo` is a type, `Foo(x)::Foo` holds always.
This breaks it.

Comment (YingboMa) [reply src/dualnumber.jl]:
```julia
julia> AbstractFloat(ForwardDiff.Dual(1, 2))::AbstractFloat
ERROR: TypeError: in typeassert, expected AbstractFloat, got ForwardDiff.Dual{Nothing,Float16,1}
Stacktrace:
 [1] top-level scope at REPL[37]:1
```
ForwardDiff doesn't handle it either, and there is nothing we can do about it.

Comment (oxinabox) [reply src/dualnumber.jl]: My question is: why do we need to define `Base.AbstractFloat(Dual)` in the first place?

Comment (YingboMa) [reply src/dualnumber.jl]: It makes AD work more code, so why not.

Comment (oxinabox) [reply src/dualnumber.jl]: 'k.
I have never seen `AbstractFloat` used this way in the wild.
Only `float`.

**Pull request merged**